### PR TITLE
Fix downward api for pod using host network.

### LINKED ISSUE #24576: When using host network, the pod IP is always empty in downward API
Turns our that when running a container in host network, its `podStatus *kubecontainer.PodStatus` that is passed to dockertools.SyncPod() will always contain empty IP, as a result, the `PodIP` environment becomes empty.

Maybe we want to use the `apiPodStatus` instead of `podStatus` in this case to [determine the pod IP](https://github.com/kubernetes/kubernetes/blob/v1.3.0-alpha.2/pkg/kubelet/dockertools/manager.go#L1805-L1808) ? Since the apiPodStatus has the valid PodIP, which is [populated by kubelet](https://github.com/kubernetes/kubernetes/blob/v1.3.0-alpha.2/pkg/kubelet/kubelet.go#L3271-L3278) when the container runs in host network.

Or just populate the podIP with hostIP in `dockertools.syncPod` when the container runs on host network.

Ref https://github.com/kubernetes/kubernetes/pull/22666

cc @aaronlevy @kubernetes/sig-node @pmorie

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes #24576

@yujuhong @yifan-gu 
I still can't find a cleaner way to fix this, so send this for discussion.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24633)

### CODE REVIEW

**Comment by dchen1107** [comment]
The code here is weird, not yours, the legacy one. :-)

Why not remove generateAPIPodStatus, instead of passing the internal PodStatus for PodIP information; When api.PodStatus is needed, call convertStatusToAPIStatus()?

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1731,6 +1731,8 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, podStatus *kubecont
 	}
 
 	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus)
+	// The pod IP may be changed by api pod status if the pod is using host network. (See #24576)
```
Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @Random-Liu I don't understand why this should ever change -- can you clarify?

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @Random-Liu see my comment: https://github.com/kubernetes/kubernetes/issues/24576#issuecomment-214770833

Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: @pmorie I mean the pod ip will be set to host ip in `generateAPIPodStatus`. Sorry for replying so late, we have meetings a whole day. :)

**Comment by aaronlevy** [comment]
@Random-Liu @pmorie @yujuhong short of plumbing the pod spec through `GetPodStatus` (https://github.com/kubernetes/kubernetes/issues/24576#issuecomment-216390335) would the change proposed here be acceptable in the interim?

**Comment by yujuhong** [comment]
@Random-Liu could you rebase and add a TODO comment to fix this properly later? Thanks.

**Comment by yujuhong** [comment]
Let's get the temporary fix in first.

**Comment by Random-Liu** [comment]
@yujuhong Just saw this. Will do!

**Comment by Random-Liu** [comment]
@yujuhong Rebased.

**Comment by pmorie** [comment]
@Random-Liu @yujuhong I think we should make a follow-up issue to fix this the right way.

**Comment by yifan-gu** [comment]
@pmorie Reopened https://github.com/kubernetes/kubernetes/issues/24576#issuecomment-219511516
But moved that issue to p2

**Comment by pmorie** [comment]
Thanks @yifan-gu

**Pull request merged**